various-receipes/typescript/test/typescript-example.ts

```diff
@@ -0,0 +1,257 @@
+        status: 200,
+        body: {
+          mode: 'approved',
+        },
+      });
+    });
+
+    // ️️️✅ Best Practice: Check the new state
+    test('When adding a new valid order, Then should be able to retrieve it', async () => {
+      //Arrange
+      const orderToAdd = {
+        userId: 1,
+        productId: 2,
+        mode: 'approved',
+      };
+
+      //Act
+      const {
+        body: { id: addedOrderId },
+      } = await request(expressApp).post('/order').send(orderToAdd);
+
+      //Assert
+      const { body, status } = await request(expressApp).get(
+        `/order/${addedOrderId}`
+      );
+
+      expect({
+        body,
+        status,
+      }).toMatchObject({
+        status: 200,
+        body: {
+          id: addedOrderId,
+          userId: 1,
+          productId: 2,
+        },
+      });
+    });
+
+    // ️️️✅ Best Practice: Check external calls
+    test('When adding a new valid order, Then an email should be send to admin', async () => {
+      //Arrange
+      process.env.SEND_MAILS = 'true';
+
+      // ️️️✅ Best Practice: Intercept requests for 3rd party services to eliminate undesired side effects like emails or SMS
+      // ️️️✅ Best Practice: Specify the body when you need to make sure you call the 3rd party service as expected
+      let emailPayload;
+      nock('http://mailer.com')
+        .post('/send', (payload) => ((emailPayload = payload), true))
+        .reply(202);
+
+      const orderToAdd = {
+        userId: 1,
+        productId: 2,
+        mode: 'approved',
+      };
+
+      //Act
+      await request(expressApp).post('/order').send(orderToAdd);
+
+      //Assert
+      // ️️️✅ Best Practice: Assert that the app called the mailer service appropriately
+      expect(emailPayload).toMatchObject({
+        subject: expect.any(String),
+        body: expect.any(String),
+        recipientAddress: expect.stringMatching(
+          /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/
+        ),
+      });
+    });
+
+    // ️️️✅ Best Practice: Check invalid input
+    test('When adding an order without specifying product, stop and return 400', async () => {
+      //Arrange
+      const orderToAdd = {
+        userId: 1,
+        mode: 'draft',
+      };
+
+      //Act
+      const orderAddResult = await request(expressApp)
+        .post('/order')
+        .send(orderToAdd);
+
+      //Assert
+      expect(orderAddResult.status).toBe(400);
+    });
+
+    // ️️️✅ Best Practice: Check error handling
+    test.todo('When a new order failed, an invalid-order error was handled');
+
+    // ️️️✅ Best Practice: Check monitoring metrics
+    test.todo(
+      'When a new valid order was added, then order-added metric was fired'
+    );
+
+    // ️️️✅ Best Practice: Simulate external failures
+    test.todo(
+      'When the user service is down, then order is still added successfully'
+    );
+
+    test('When the user does not exist, return 404 response', async () => {
```

Comment:
Should we copy all the tests also they look the same? Maybe just leave a link to the origin js file?